### PR TITLE
PPM-319 Fix DCS scan timeout on no results

### PR DESCRIPTION
**Pull request by itayx**
Currently the NL messages are received for both interfaces regardless
of which started the scan.
This becomes an issue with the Channel_Scan_Finished message, since it's
only identifier is the message's sequence number (nlh->nlmsg_seq).
So to solve this issue we store the sequence number present in the
Channel_Scan_Dump_Result message, and validate it against the one
received in the Channel_Scan_Finished message.

But when the scan returns with no neighboring APs the scan dump sequence
only contains the Channel_Scan_Finished, so we don't have an initial
sequence number to store, and later on validate against.

Added a scan dump in progress flag and validate against it, since the
DCS task does not allow for two scans in progress.

PPM-319
Signed-off-by: Itay Elenzweig <itayx.elenzweig@intel.com>

### CODE REVIEW

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h [hunk 1]
```diff
@@ -113,7 +113,8 @@ class mon_wlan_hal_dwpal : public base_wlan_hal_dwpal, public mon_wlan_hal {
     }
 
     std::shared_ptr<char> m_temp_dwpal_value;
-    uint32_t m_nl_seq = 0;
+    uint32_t m_nl_seq            = 0;
```
Comment (morantr) [inline common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h:116]: Can you please add a comment explaining this member? Its name is not so self-explanatory.

Comment (morantr) [inline common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h]: Can you please add a comment explaining this member? Its name is not so self-explanatory.

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp [hunk 2]
```diff
@@ -1389,26 +1386,27 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
...
         event_queue_push(event);
         break;
     }
     case Event::Channel_Scan_Finished: {
-        // ifname is corrupted for Channel_Scan_Finished event using nlh->nlmsg_seq instead.
-        if (nlh->nlmsg_seq != m_nl_seq) {
-            // Current event has a sequence number not matching the current sequence number
+        // ifname is invalid for Channel_Scan_Finished event using nlh->nlmsg_seq instead.
+        // checking nlh->nlmsg_seq does not work when there are no neighboring results.
+        if (/*nlh->nlmsg_seq != m_nl_seq*/ !m_scan_dump_in_progress) {
```
Comment (morantr) [inline common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: Why did you leave the comment on the condition?
Also, it does not fit to the comment above, because you don't check `nlh->nlmsg_seq`.

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp [hunk 3]
```diff
@@ -1346,21 +1346,17 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
             return true;
         }
 
-        // We need to distinguish 1st and rest of dump events (NL80211_CMD_NEW_SCAN_RESULTS)
-        // 1st event is an empty dump result that we should reply with scan dump request.
-        // nlh->nlmsg_seq = 0 only with the 1st dump result.
-        // rest of events are the actual scan dump results that need to be parsed.
-        // unique sequence number is chosen by the nl (nlh->nlmsg_seq != 0) for the rest of events.
+        // Validate dump sequence number is cleared & no scan is in progress
         if (m_nl_seq == 0) {
```
Comment (morantr) [inline common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: I think you should keep the comment "We need to distinguish...".
It helps to understand the flow.
You can add the new comment after it.

You are saying on the new comment that you validate that "no scan is in progress" but I don't see it being validated.

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp [hunk 4]
```diff
@@ -1346,21 +1346,17 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
...
                 event_queue_push(Event::Channel_Scan_New_Results_Ready);
+                m_scan_dump_in_progress = true;
                 channel_scan_dump_results();
                 return true;
-            } else {
-                LOG(DEBUG) << "Results dump new sequence:" << int(nlh->nlmsg_seq);
-                m_nl_seq = nlh->nlmsg_seq;
             }
+            LOG(DEBUG) << "Results dump new sequence:" << int(nlh->nlmsg_seq);
+            m_nl_seq = nlh->nlmsg_seq;
```
Comment (morantr) [inline common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: You got rid of nesting! Good! :)

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp [hunk 5]
```diff
@@ -1374,6 +1370,7 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
             }
             LOG(DEBUG) << "Processing results for BSSID:" << results->channel_scan_results.bssid;
             event_queue_push(event, results);
+            return true;
         } else {
             LOG(ERROR) << "channel scan results dump received with unexpected seq number";
             return false;
```
Comment (morantr) [inline common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: You can get rid of the nesting here too, although I'm not sure I understand why did you add the`return true` here.

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp (hunk 3 above)
Comment (kantera800) [reply common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: IMO you should keep the comment above

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp [hunk 6]
```diff
@@ -1389,26 +1386,27 @@ bool mon_wlan_hal_dwpal::process_dwpal_nl_event(struct nl_msg *msg)
...
+        m_nl_seq                = 0;
+        m_scan_dump_in_progress = false;
         event_queue_push(event);
         break;
     }
     case Event::Channel_Scan_Finished: {
-        // ifname is corrupted for Channel_Scan_Finished event using nlh->nlmsg_seq instead.
-        if (nlh->nlmsg_seq != m_nl_seq) {
-            // Current event has a sequence number not matching the current sequence number
+        // ifname is invalid for Channel_Scan_Finished event using nlh->nlmsg_seq instead.
```
Comment (kantera800) [inline common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: we should open a driver bug for the

>  ifname is invalid for Channel_Scan_Finished

**Review by kantera800 — APPROVED** [review]

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp (hunk 3 above)
Comment (itayx) [reply common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: I re-wrote the comment, with better grammar, and a clearer explanation to the scan dump results process

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp (hunk 5 above)
Comment (itayx) [reply common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: Because of the nesting I wanted the correct result to return `true`
So i removed the nesting, and now it goes to `break` which returns `true `

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h (hunk 1 above)
Comment (itayx) [reply common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.h]: Added a comment for `m_nl_seq` and `m_scan_dump_in_progress`

File: common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp (hunk 2 above)
Comment (itayx) [reply common/beerocks/bwl/dwpal/mon_wlan_hal_dwpal.cpp]: Rewrote the `Channel_Scan_Finished` handling
We still check the `nlh->nlmsg_seq` but only if the `m_nl_seq` is set (meaning the sequence number is valid)

**Review by morantr — APPROVED** [review]
